packages/elements/src/internals/machines/sign-in/router.selectors.ts

```diff
@@ -8,18 +8,24 @@ export function SignInSafeIdentifierSelectorForStrategy(
 ): (s: SignInRouterSnapshot) => string {
   return (s: SignInRouterSnapshot) => {
     const signIn = s.context.clerk?.client.signIn;
-    const identifier = signIn.identifier || '';
 
     if (strategy) {
-      const matchingFactor = [...(signIn.supportedFirstFactors ?? []), ...(signIn.supportedSecondFactors ?? [])].find(
-        f => f.strategy === strategy,
-      );
+      const matchingFactors = [
+        ...(signIn.supportedFirstFactors ?? []),
+        ...(signIn.supportedSecondFactors ?? []),
+      ].filter(f => f.strategy === strategy);
+
+      const matchingFactor =
+        signIn.identifier && matchingFactors.length > 0
+          ? matchingFactors.find(f => 'safeIdentifier' in f && f.safeIdentifier === signIn.identifier)
```

Comment:
Requiring the factor's safeIdentifier to match the user provided identifier results in the user provided identifier being returned for factors in which it's not valid, for example rendering a phone number for `email_code` and `email_link`. Should we maybe break this logic up into two pieces?

```ts
const matchingFactorForIdentifier =
  signIn.identifier && matchingFactors.length > 0
    ? matchingFactors.find(
        (f) => "safeIdentifier" in f && f.safeIdentifier === signIn.identifier
      )
    : null;

const matchingFactorForStrategy = matchingFactors[0];

if (
  matchingFactorForIdentifier &&
  "safeIdentifier" in matchingFactorForIdentifier
) {
  return matchingFactorForIdentifier.safeIdentifier;
} else if (
  matchingFactorForStrategy &&
  "safeIdentifier" in matchingFactorForStrategy
) {
  return matchingFactorForStrategy.safeIdentifier;
}
```

![Screen Shot 2024-08-16 at 10 37 18 AM](https://github.com/user-attachments/assets/e3b62435-563c-49b6-8ce4-7585af4c44aa)